### PR TITLE
Add artifact.test.* messages to pipeline via contra-lib closures

### PR DESCRIPTION
**Pull request by johnbieren**
I left the existing messages in place as I don't really know how all the backend stuff is hooked up, so I am not sure when they can be removed in favor of the new messages (if ever). Once the PR above is merged and we can test this, I'd ask the reviewers to actually look at datagrepper and make sure all the messages are as expected.

We will need a PR to fedmsg_meta for the new topics. I can tackle that once the contra-lib PR is merged and I have some stage messages to go from.

I will raise a future PR with the pipeline.running, pipeline.complete, and pipeline.stage messages. I haven't added those to the contra-lib yet.

Finally, I am commenting in line with some things I'd like double check that I arbitrarily chose.

### CODE REVIEW

File: src/org/centos/pipeline/PackagePipelineUtils.groovy [hunk 1]
```diff
@@ -91,6 +91,63 @@ def setMessageFields(String messageType, String artifact) {
...
+ * to create the ci.artifact.test.messageType messages
+ * @param messageType: queued, running, complete, error
+ * @param artifact: dist-git-pr, koji-build
+ * @return
+ */
+// Plan is to rename to setMessageFields and remove function above once everything seems fine and stable
+def setTestMessageFields(String messageType, String artifact) {
+    // Set values that go in multiple closures
+    myTopic = "${MAIN_TOPIC}.ci.${artifact}.test.${messageType}"
+    myType = (artifact == 'koji-build') ? 'tier0' : 'build'
```
Comment (johnbieren) [inline src/org/centos/pipeline/PackagePipelineUtils.groovy]: @thrix please ack/nack the MESSAGE_TYPES I chose

File: src/org/centos/pipeline/PackagePipelineUtils.groovy [hunk 2]
```diff
@@ -91,6 +91,63 @@ def setMessageFields(String messageType, String artifact) {
...
+    myComponent = env.fed_repo
+    myIssuer = env.fed_owner ?: fed_username
+    // PR builds are scratch so defaulting to scratch = true
+    myScratch = env.isScratch ? env.isScratch.toBoolean() : true
+    taskid = env.fed_task_id ?: env.fed_id
+    myId = (artifact == 'koji-build') ? taskid : env.fed_rev
```
Comment (johnbieren) [inline src/org/centos/pipeline/PackagePipelineUtils.groovy]: @thrix please ack/nack using the fed_rev for the id field in the case of a dist-git-pr (this goes in the header and artifact closures)

File: src/org/centos/pipeline/PackagePipelineUtils.groovy [hunk 3]
```diff
@@ -91,6 +91,63 @@ def setMessageFields(String messageType, String artifact) {
...
+    myNvr = env.nvr ?: ""
+    myCategory = (artifact == 'koji-build') ? 'functional' : 'static-analysis'
```
Comment (johnbieren) [inline src/org/centos/pipeline/PackagePipelineUtils.groovy]: @thrix please ack/nack the category. I got these looking at other examples, but we can obviously change them.

File: src/org/centos/pipeline/PackagePipelineUtils.groovy [hunk 4]
```diff
@@ -91,6 +91,63 @@ def setMessageFields(String messageType, String artifact) {
...
+    myNamespace = "fedora-ci"
```
Comment (johnbieren) [inline src/org/centos/pipeline/PackagePipelineUtils.groovy]: @thrix @thrix please ack/nack the namespace value I chose (goes in the root level of all the test.* messages)

File: src/org/centos/pipeline/PackagePipelineUtils.groovy [hunk 5]
```diff
@@ -91,6 +91,63 @@ def setMessageFields(String messageType, String artifact) {
...
+
+    // Create message header
+    myHeader = msgBusHeader(topic: myTopic, type: myType, component: myComponent, issuer: myIssuer, scratch: myScratch, id: myId, nvr: myNvr)
+    // Create common message body content
+    myCIContent = msgBusCIContent(name: "fedora-ci", team: "fedora-ci", irc: "#fedora-ci", email: "ci@lists.fedoraproject.org")
```
Comment (johnbieren) [inline src/org/centos/pipeline/PackagePipelineUtils.groovy]: @thrix finally, please ack/nack the values I chose for these fields.

**Comment by johnbieren** [comment]
[test]

**Comment by johnbieren** [comment]
[test] script approval had to be updated

**Comment by johnbieren** [comment]
[test] another script approval

**Comment by johnbieren** [comment]
[test] and another one

**Comment by johnbieren** [comment]
[test] hopefully last one

**Comment by johnbieren** [comment]
I believe the stage datagrepper process is messed up right now.

**Comment by johnbieren** [comment]
[test] fedmsg.stg is fixed

**Comment by johnbieren** [comment]
When I set fields for the header and fields for the body, I end up with this
https://apps.stg.fedoraproject.org/datagrepper/id?id=2018-f82a497e-169d-4c60-b1cf-adf30988a034&is_raw=true&size=extra-large

You see there is headers: {} .. I am going to keep investigating what is happening here. It seems to be fedmsg specific

File: src/org/centos/pipeline/PackagePipelineUtils.groovy (hunk 4 above)
Comment (thrix) [reply src/org/centos/pipeline/PackagePipelineUtils.groovy]: Yep, can be for now, we can always change later if needed

File: src/org/centos/pipeline/PackagePipelineUtils.groovy (hunk 1 above)
Comment (thrix) [reply src/org/centos/pipeline/PackagePipelineUtils.groovy]: Yeah, I think this is ok for Fedora pipeline. with dist-git-pr artifact we choose build, yep

File: src/org/centos/pipeline/PackagePipelineUtils.groovy (hunk 3 above)
Comment (thrix) [reply src/org/centos/pipeline/PackagePipelineUtils.groovy]: This is ok for the pipeline

**Comment by thrix** [comment]
@johnbieren thanks for working on this. Is there something we could help with in regards to:

> You see there is headers: {} .. I am going to keep investigating what is happening here. It seems to be fedmsg specific

Maybe we could start using the new AMQP fedmsg2 instead? So we do not deal with the old legacy zero MQ based specifics?

**Comment by johnbieren** [comment]
@thrix thank you for addressing my concerns with the data values I guessed at. As far as the amqp vs zero mq thing, I am willing to try it. I am not sure if it is a Jenkins plugin issue or the different kinds of bus. The issue is that the json wrapping like overdoes it as opposed to when using the UMB. So, it might just be the way the jms-messaging plugin handles message bodies when prepping for fedmsg. I am still actively investigating it (just have been also spending time with some other message stuff for internal pipelines).

File: src/org/centos/pipeline/PackagePipelineUtils.groovy (hunk 2 above)
Comment (thrix) [reply src/org/centos/pipeline/PackagePipelineUtils.groovy]: @johnbieren what is fed_rev please?

Comment (johnbieren) [reply src/org/centos/pipeline/PackagePipelineUtils.groovy]: Either the koji task id or "PR-X" where X is the pr number. We can change that if something else fits better there

Comment (thrix) [reply src/org/centos/pipeline/PackagePipelineUtils.groovy]: I do not think that is sufficient the artifact ID should be unique. What about using pull request id instead?

Comment (johnbieren) [reply src/org/centos/pipeline/PackagePipelineUtils.groovy]: Sure thing, I'll push an update soon

Comment (johnbieren) [reply src/org/centos/pipeline/PackagePipelineUtils.groovy]: @thrix the pull request uid? The id is only unique per repo.. unless that is what you meant by unique

**Comment by johnbieren** [comment]
So resultsDB updater will have to be updated to receive these messages. Probably need a change for fedmsg_meta for the new topics being used by CentOS CI as well. The old messages are still in the Jenkinsfile, but I am hoping we can remove those shortly after merging this?

**Comment by johnbieren** [comment]
running - https://apps.stg.fedoraproject.org/datagrepper/id?id=2019-013e0cd3-2964-4dfe-b34e-39234e49cfed&is_raw=true&size=extra-large
error (because it is failing in nvr-verify stage) - https://apps.stg.fedoraproject.org/datagrepper/id?id=2019-3d475c5f-5ca7-4598-afc3-23f155c2f2a2&is_raw=true&size=extra-large

**Comment by AdamWill** [comment]
"So resultsDB updater will have to be updated to receive these messages. Probably need a change for fedmsg_meta for the new topics being used by CentOS CI as well. The old messages are still in the Jenkinsfile, but I am hoping we can remove those shortly after merging this?"

it's not *required* to stop publishing 'native' messages, FWIW - it's perfectly OK to do both. openQA publishes both 'standardized' and 'native' messages too. 

It would probably be a good idea to publish to AMQP ([https://github.com/fedora-infra/fedora-messaging](fedora-messaging)) if you can manage that, indeed. @jeremycline is trying to deprecate fedmsg (older 0MQ thing) now, so it would be bad to start publishing new fedmsgs if we can avoid it. All the [docs](https://fedora-messaging.readthedocs.io/en/stable/) and stuff for fedora-messaging are sort of set up around publishing from Python, but it shouldn't be too hard to do it from Java instead. I'm currently working on having openQA publish to AMQP instead of 0MQ as well.

**Comment by jeremycline** [comment]
Yeah, any AMQP 0.9.1 client is fine, the message format is documented at https://fedora-messaging.readthedocs.io/en/stable/wire-format.html

**Comment by AdamWill** [comment]
I came up with [a thing](https://pagure.io/fedora-qa/python-ci_messages/) today which wraps the CI Messages schemas to the fedora-messaging message schema class system.

**Comment by johnbieren** [comment]
> LGTM, I'd just use sendMessage instead of sendMessageWithAudit as we changed this for the other topics.

Done

**Comment by bgoncalv** [comment]
Looks good to me:

https://jenkins-continuous-infra.apps.ci.centos.org/view/Fedora%20All%20Packages%20Pipeline/job/fedora-rawhide-stage-pr-pipeline/403/consoleText

org.centos.stage.ci.dist-git-pr.test.running

    {"msg":{"artifact":{"comment_id":0,"commit_hash":"1c474a3f55b792c378c6f1a0a8075335d87eff03","id":"6","issuer":"bgoncalv","repository":"https://src.fedoraproject.org/rpms/ksh","source":"UNKNOWN","type":"pull-request","uid":"b08cc62005f643a39fa66f7c4891dac4"},"contact":{"docs":"https://pagure.io/standard-test-roles","email":"ci@lists.fedoraproject.org","irc":"#fedora-ci","name":"fedora-ci","team":"fedora-ci","url":"https://jenkins-continuous-infra.apps.ci.centos.org/"},"generated_at":"2019-08-23T14:50:27.346Z","pipeline":{"build":"403","id":"51310ab3-5391-425c-b4ac-4871b79cb9d0","name":"fedora-rawhide-stage-pr-pipeline","stage":{"name":"prepare-environment"}},"run":{"log":"https://jenkins-continuous-infra.apps.ci.centos.org/job/fedora-rawhide-stage-pr-pipeline/403/console","rebuild":"https://jenkins-continuous-infra.apps.ci.centos.org/job/fedora-rawhide-stage-pr-pipeline/403/rebuild/parameterized","url":"https://jenkins-continuous-infra.apps.ci.centos.org/blue/organizations/jenkins/fedora-rawhide-stage-pr-pipeline/detail/fedora-rawhide-stage-pr-pipeline/403/pipeline/"},"test":{"category":"static-analysis","namespace":"fedora-ci.dist-git-pr","type":"build"},"version":"0.2.1"},"msg_id":"2019-e3a6f119-32b9-4504-b60e-3340620bfeea","timestamp":0,"topic":"org.centos.stage.ci.dist-git-pr.test.running"}

org.centos.stage.ci.dist-git-pr.test.complete

    {"msg":{"artifact":{"comment_id":0,"commit_hash":"1c474a3f55b792c378c6f1a0a8075335d87eff03","dependencies":[],"id":"6","issuer":"bgoncalv","repository":"https://src.fedoraproject.org/rpms/ksh","source":"UNKNOWN","type":"pull-request","uid":"b08cc62005f643a39fa66f7c4891dac4"},"contact":{"docs":"https://pagure.io/standard-test-roles","email":"ci@lists.fedoraproject.org","irc":"#fedora-ci","name":"fedora-ci","team":"fedora-ci","url":"https://jenkins-continuous-infra.apps.ci.centos.org/"},"generated_at":"2019-08-23T15:00:27.725Z","pipeline":{"build":"403","id":"51310ab3-5391-425c-b4ac-4871b79cb9d0","name":"fedora-rawhide-stage-pr-pipeline","stage":{"name":"package-tests"}},"run":{"log":"https://jenkins-continuous-infra.apps.ci.centos.org/job/fedora-rawhide-stage-pr-pipeline/403/console","rebuild":"https://jenkins-continuous-infra.apps.ci.centos.org/job/fedora-rawhide-stage-pr-pipeline/403/rebuild/parameterized","url":"https://jenkins-continuous-infra.apps.ci.centos.org/blue/organizations/jenkins/fedora-rawhide-stage-pr-pipeline/detail/fedora-rawhide-stage-pr-pipeline/403/pipeline/"},"system":[{"architecture":"x86_64","label":"upstream-fedora-pipeline","os":"master","provider":"CentOS CI","variant":"Cloud"}],"test":{"category":"static-analysis","namespace":"fedora-ci.dist-git-pr","result":"passed","type":"build"},"version":"0.2.1"},"msg_id":"2019-e1b54e70-bfc4-48e9-85bf-e64b5328e7b9","timestamp":0,"topic":"org.centos.stage.ci.dist-git-pr.test.complete"}

**Comment by bgoncalv** [comment]
@thrix are we good to merge this?

**Comment by thrix** [comment]
@johnbieren som last nitpicks:

```
    "artifact": {
      "comment_id": 0,
      "commit_hash": "1c474a3f55b792c378c6f1a0a8075335d87eff03",
      "id": "6",
      "issuer": "bgoncalv",
      "repository": "https://src.fedoraproject.org/rpms/ksh",
      "source": "UNKNOWN",
      "type": "pull-request",
      "uid": "b08cc62005f643a39fa66f7c4891dac4"
    },
    ....
     "timestamp": 0,
```
1. The comment_id should be ideally a string, seems we are using strins only elsewhere ...

2. source field should not be there ...

3.  "type" should be "dist-git-pr", this is incorrect also in pagure.io/fedora-ci/messages ....

4. timestamp is incorrect

5. in complete message I also see:

```
      "dependencies": [],
```

**Comment by bgoncalv** [comment]
>     4. timestamp is incorrect

the timestamp issue I think we had an issue reported about it, but I could not find it. From what I can remember this is an issue with the message plugin. Maybe it is fixed with the newer version of the plugin.

**Comment by thrix** [comment]
All might spec is right, I am sorry @johnbieren !

**Comment by thrix** [comment]
> All might spec is right, I am sorry @johnbieren !

Ah, it is not there, but it must be, so I think it is fine

**Pull request merged**